hadoop-hdds/docs/content/design/full-volume-handling.md

```diff
@@ -0,0 +1,166 @@
+---
+title: Full Volume Handling
+summary: Immediately trigger Datanode heartbeat on detecting full volume
+date: 2025-05-12
+jira: HDDS-12929
+status: Design 
+author: Siddhant Sangwan, Sumit Agrawal
+---
+
+<!--
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License. See accompanying LICENSE file.
+-->
+
+## Summary
+On detecting a full Datanode volume during write, immediately trigger a heartbeat containing the latest storage 
+report for all volumes.
+
+## Problem
+When a Datanode volume is close to full, the SCM may not be immediately aware of this because storage reports are only
+sent to it every one minute (`HDDS_NODE_REPORT_INTERVAL_DEFAULT = "60s"`). We would like SCM to know about this as 
+soon as possible, so it can make an informed decision when checking the volumes in that Datanode for deciding whether a 
+new pipeline can contain that Datanode (in `SCMCommonPlacementPolicy.hasEnoughSpace`).
+
+Additionally, SCM only has stale
+information about the current size of a container because container size is only updated when an Incremental Container
+Report (event based, for example when a container transitions from open to closing state) is received or a Full
+Container Report (`HDDS_CONTAINER_REPORT_INTERVAL_DEFAULT = "60m"`) is received. This can lead to the SCM
+over-allocating blocks to containers on a full DN volume. When the writes eventually fail, performance will drop
+because the client will have to request for a different set of blocks. We will discuss how we tried to solve this, 
+but ultimately decided to not go ahead with the solution.
```

Comment:
I'm confused on the status of this doc. What part of it maps to what we actually plan to implement?

Reply:
The design changed a lot. We've already merged the corresponding implementation - https://github.com/apache/ozone/pull/8590, and I've updated this doc now. 

This comment will tell you why and what changed - https://github.com/apache/ozone/pull/8492#issuecomment-2954967269

This comment and the design doc itself will tell you what we implemented - https://github.com/apache/ozone/pull/8590#issue-3129500113

In short, we trigger heartbeat immediately including the close container action. Per container throttling. Not sending storage reports.